### PR TITLE
metal: 1-bit qmv_fast use 1 pack/thread for occupancy (~9% on M5)

### PR DESCRIPTION
**Pull request by bri-prism**
## What

`affine_qmv_fast` (the decode/token-gen quantized matvec) sets
`packs_per_thread = bits == 2 ? 1 : 2`, so **1-bit** falls into the `: 2` case and gets
`values_per_thread = pack_factor * 2 = 64` → `x_thread[64]` (~256 B/thread). That register
footprint drops occupancy, so 1-bit decode can't hide DRAM latency and **under-saturates
bandwidth (~75% on M5 Pro)** while 2-bit/4-bit saturate (~90/96%).

Use **1 pack/thread for `bits <= 2`** so 1-bit matches 2-bit's `values_per_thread = 32`
footprint. (2-bit already used 1 pack, so it is unchanged — it serves as a control below.)

## Why it helps

- **Occupancy:** halves per-thread register/`x_thread` pressure for 1-bit → better latency hiding.
- **Bonus correctness/robustness:** `scale_step_per_thread = group_size / values_per_thread` was
  `0` for `group_size = 32` at 1-bit (vpt=64); with vpt=32 it is well-defined for g32.

## Measured (Apple M5 Pro, distinct-weight DRAM-bound matvec, best-of-10)

2-bit is the drift control (untouched by this change):

| | before (packs=2) | after (packs=1) | Δ |
|---|---|---|---|
| **1-bit** | 24.0 µs/matvec | **21.9 µs/matvec** | **~9% faster** (75%→82% BW) |
| 2-bit (control) | 33.2 µs | 32.9 µs | 0.8% (drift) |

Correct: `quantized_matmul` vs dequantize+matmul rel_err **2.7e-4**. The control moving only 0.8%
confirms the 1-bit gain is real, not machine noise.

## Notes

- Measured on a contended M5 (shared with another profiling session) — the ~9% is robust vs the
  control, but a quiet machine would tighten the headline number.
- Remaining 1-bit under-saturation (82%→~95%) is dominated by scale+bias traffic (~33% of 1-bit
  bytes at group 64) — a follow-up (int8/shared scales), not this PR.

### CODE REVIEW

**Comment by bri-prism** [comment]
End-to-end validation on the test 1-bit model (MLX, M5 Pro; 128 tok, temp 0, 3 reps each):

| | tokens/sec |
|---|---|
| before (packs=2) | 31.0 (30.97/30.99/31.03) |
| after (packs=1) | **34.2** (34.13/34.25/34.32) |
| **decode speedup** | **+10.3%** |

Both stable across reps. The occupancy fix lifts every projection matvec in the decode path, so the end-to-end gain slightly exceeds the isolated qmv micro-bench (~9%).

**Comment by bri-prism** [comment]
Context-window scaling (test 1-bit model, M5 Pro, decode t/s after prompt prefill):

| context (prompt tok) | before | after | win |
|---|---|---|---|
| 426 | 30.8 | 34.1 | +10.7% |
| 1,650 | 30.0 | 33.1 | +10.6% |
| 6,570 | 27.5 | 30.2 | +9.8% |
| 13,122 | 24.8 | 27.0 | +8.7% |

The win persists across context, diluting only slightly (10.7%→8.7%) as the attention KV cache grows to a larger share of decode (the hybrid-attention design keeps that share small — decode falls only ~19% over 30× context).

**Comment by bri-prism** [comment]
Losslessness confirmed — **bit-exact** on the test 1-bit model. Teacher-forced prefill over a fixed passage (33 positions × full vocab), before vs after:

```
mean KLD: 0.000e+00   max KLD: 0.000e+00
top-1 agreement: 100.00% (33/33)
logit |Δ|: mean 0.0, max 0.0   ->  arrays identical
```

The occupancy change (vpt 64→32) reorders the K-loop tiling but yields byte-identical logits, so output is unchanged while decode is +10%. (Note: the 1-bit checkpoint itself currently generates incoherent text — a separate, pre-existing conversion-coherence issue unrelated to this kernel; identical before/after.)

**Review by khosravipasha — APPROVED** [review]

**Comment by khosravipasha** [comment]
On m4 got 20% speed up so seems great :)
